### PR TITLE
ZSTD 1.4.6

### PR DESCRIPTION
**Pull request by Ornias1993**
### Motivation, Context and Description
This PR introduces ZSTD 1.4.6 Staging (ZSTD 1.4.6 as available in their staging branch) and tests available new flags.

### Note
Can not be added yet, wait for final release and merge of zstd 1.4.6!

### How Has This Been Tested?
It has passed the test suit 100% on all systems.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by Ornias1993** [comment]
@terrelln 
Considering most of 1.4.6 of ZSTD is (going to be) your work on kernel support (which also features some flags we can using) some questions:
- Should or shouldn't we use `ZSTD_NO_INTRINSICS` (and what are the consequences of using it in practice?)
- Should or shouldn't we set `ZSTD_COMPRESS_HEAPMODE` once merged? (don't think it has relevance with our custom allocator, does it?)
- In https://github.com/facebook/zstd/commit/978659d34f4e0ec1bb2a1e2eebb137b925f4ca43 you void some memory functions, should we also do so, considering we have a custom zstd_deps.h and use a custom allocator?
- Are there any more compile flags we should be using?

**Comment by terrelln** [comment]
> `ZSTD_NO_INTRINSICS`

Zstd in the kernel didn't compile without this enabled, because the headers weren't available. I would expect no speed difference on x86, and maybe slightly negative on aarch64.

> `ZSTD_COMPRESS_HEAPMODE`

This only effects `ZSTD_compress()` which you don't use, since it doesn't accept a custom allocator. It is enabled in the kernel to reduce the stack usage of `ZSTD_compress()`, so the build doesn't warn about large frame sizes.

> Are there any more compile flags we should be using?

Just `ZSTD_NO_UNUSED_FUNCTIONS` if you want to avoid some of the functions that use a ton of stack space, but are unused.

**Comment by Ornias1993** [comment]
@terrelln Thanks for the quick response.
`ZSTD_NO_UNUSED_FUNCTIONS` is indeed a godsend, thank you! :)

**Comment by Ornias1993** [comment]
Note: Basically ready, just needs an update on release of ZSTD 1.4.6

**Comment by terrelln** [comment]
Also, you may need to set `ZSTD_ADDRESS_SANITIZER=0` and `ZSTD_MEMORY_SANITIZER=0`, unless you disable that code in another way. In the Linux Kernel, zstd detected ASAN support, and tried to call into the ASAN library. But, since it is KASAN, the same runtime library support wasn't available. I don't know if you have that problem here, but it is worth pointing out.

**Comment by Ornias1993** [comment]
@terrelln any potential downsides to doing so?

**Comment by IvanVolosyuk** [comment]
I wonder what's the story about compression stability for the dedup purposes? Just out of curiosity, this might not apply for pre-release versions of ZFS.

**Comment by Ornias1993** [comment]
@IvanVolosyuk I'll point you to the mailinglist and previous PR's to discuss this.
This PR is not the place and I do not want a general ZSTD-on-ZFS discussion in my attempt to get an update going.

**Comment by terrelln** [comment]
> @terrelln any potential downsides to doing so?

Nope! You just lose some extra poisoning that we do when MSAN/ASAN is enabled (if it does work). But the value of that is really for our own testing / fuzzers, I don't think it would expose any mis-usages of the API.

**Comment by terrelln** [comment]
> I wonder what's the story about compression stability for the dedup purposes?

Zstd guarantees reproducibility on the same version with the same options, but guarantees no output stability between different zstd versions, and the compressed output is often (slightly) different between versions. Though of course it is always both forwards and backwards comparable.

**Comment by darkfiberiru** [comment]
@IvanVolosyuk ZFS uses hash of compressed block not block before compression? As long as it using hash of block prior to compression all we need is compatibility to decompress block between zstd version which is something zstd promises.

**Comment by Ornias1993** [comment]
@darkfiberiru It doesn't, according to another thread.
But not relevant to this PR, so again: Please create a mailinglist topic or something.

**Comment by Ornias1993** [comment]
@terrelln Not trying to be a bother, but how is the progress going on 1.4.6 (and more precisely: Kernel Acceptance)

File: module/zstd/README.md [hunk 1]
```diff
@@ -29,30 +29,18 @@ To update ZSTD the following steps need to be taken:
...
-    "${url}/releases/download/v${release}/zstd-${release}.tar.gz"
-tar -C /tmp -xzf /tmp/zstd.tar.gz
-
-cp ${zstd}/lib/zstd.h module/zstd/lib/
-cp ${zstd}/lib/zstd_errors.h module/zstd/lib/
-${zstd}/contrib/single_file_libs/combine.sh \
-    -r ${zstd}/lib -o module/zstd/lib/zstd.c module/zstd/zstd-in.c
-~~~
+This can be done using a few shell commands from inside the zfs repo.
+"update_zstd.sh" is included as a scripted example of how to update the ZSTD library.
```
Comment (c0d3z3r0) [inline module/zstd/README.md:33]: I'd do that in a separate commit / PR

File: module/zstd/README.md [hunk 2]
```diff
@@ -29,30 +29,18 @@ To update ZSTD the following steps need to be taken:
...
 the macro list in include/zstd_compat_wrapper.h usually needs to be updated.
-this can be done with some hand crafting of the output of the following
-script: nm zstd.o | awk '{print "#define "$3 " zfs_" $3}' > macrotable
+One can use the following script to create a list of required macros: 
+```
+nm zstd.o | awk '{print "#define "$3 " zfs_" $3}' > macrotable
+```
 
+This list then needs to be manually filtered for relevance, validity, duplicates, blank lines etc.
+_TODO: Add a script to automate some repeatitive bits of this a little more_
```
Comment (c0d3z3r0) [inline module/zstd/README.md:43]: `nm -fposix zstd.o | awk '{print "#define "$1 " zfs_" $1}' |  sort -u > macrotable`

^ this saves you from "empty lines" (I guess you mean such: "#define  zfs_") and drops duplicates

File: module/zstd/zstd-in.c [hunk 3]
```diff
@@ -41,6 +41,16 @@
 #define	ZSTD_LIB_DICTBUILDER 0
 #define	ZSTD_LIB_DEPRECATED 0
 #define	ZSTD_NOBENCH
+#define	ZSTD_NODICT
+#define	ZSTD_NO_INTRINSICS
+#define	ZSTD_NO_UNUSED_FUNCTIONS
```
Comment (c0d3z3r0) [inline module/zstd/zstd-in.c:46]: with this being added you can try to remove the stack-warning workarounds (Wframe-larger-than=20480) in `module/zstd/Makefile.in` and `lib/libzstd/Makefile.am`

File: module/zstd/README.md (hunk 1 above)
Comment (Ornias1993) [reply module/zstd/README.md:33]: It's the same script thats already included.
It's just split into a seperate file.

File: module/zstd/zstd-in.c (hunk 3 above)
Comment (Ornias1993) [reply module/zstd/zstd-in.c:46]: They should already be, let me check again.

Comment (Ornias1993) [reply module/zstd/zstd-in.c:46]: Module one is indeed removed in this PR already and the lib makefile has already the stack warning removed in a earlier PR by someone else... 👍

File: module/zstd/README.md (hunk 2 above)
Comment (Ornias1993) [reply module/zstd/README.md:43]: Awesome, thanks... Didn't have the time yet to figure that out, will look into it/include it :)
(either in the .sh or the readme whatever the final choice would be)

Comment (Ornias1993) [reply module/zstd/README.md:43]: Changed, thanks!

File: module/zstd/README.md (hunk 1 above)
Comment (c0d3z3r0) [reply module/zstd/README.md:33]: well, I just felt "update zstd" and "rework/move docs" is not really the same change

Comment (Ornias1993) [reply module/zstd/README.md:33]: Not going to change it.
It needed cleaning and scripts shouldn't be in the readme. I'm not going to bother the maintainers and myself with two PR's.

**Comment by Ornias1993** [comment]
As soon as this is passing the tests, i'll clean up a little and file a new PR as it's now 1.4.7 and this WIP PR is ancient by now.

**Comment by Ornias1993** [comment]
Awesome, except style and some known failures everything checks out

**Comment by Ornias1993** [comment]
Succeeded by #11367
